### PR TITLE
PLAD-67 - As a user, I want to be able to log in, so that I have access to the system

### PR DESCRIPTION
**Pull request by Fabs97**

### CODE REVIEW

**Review by AliceCasali — APPROVED** [review]
Everything seems correct from a frontend point of view, the app is running correctly (finallyyyyyy)

**Review by AnnaBergamasco — CHANGES_REQUESTED** [review]
Overall super good job!
Most of my comments are about missing error handling parts that aren't actually that vital, but there are also three TODOs left in the code.

File: backend/DAO/employerDAO.js [hunk 1]
```diff
@@ -16,21 +16,30 @@ module.exports = {
...
-        let result = await database("employer")
-            .select("id")
-            .orderBy("id", "desc")
-            .limit(1);
-        return result[0];
+    getEmployerByUserId:  async (userId) => {
+        let result = await database('employer')
+            .select('id', 'name')
+            .where('user_id', userId)
+
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/DAO/employerDAO.js:23]: added

File: backend/DAO/employerDAO.js [hunk 2]
```diff
@@ -16,21 +16,30 @@ module.exports = {
...
+            .where('user_id', userId)
+        
+        return result.length ? result[0] : null;
+    },
+
+    getEmployerByPlacementId: async (id) => {
+        let result = await database('employer as e')
+            .select('e.id', 'e.user_id as userId')
+            .leftJoin('placements as p', 'e.id', 'p.employer_id')
+            .where('p.id', id);
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/DAO/employerDAO.js:31]: added

File: backend/DAO/matchDAO.js [hunk 3]
```diff
@@ -80,5 +80,14 @@ module.exports = {
...
                 }
             })
+    },
+    getStudentMatchesWithEmployer: async function (employerId, studentID) {
+        let result = await database('student_has_placement as shp')
+            .select('shp.student_id', 'shp.placement_id', 'shp.student_accept', 'shp.placement_accept', 'shp.status')
+            .leftJoin('placements as p', 'shp.placement_id', 'p.id')
+            .where('p.employer_id', employerId)
+            .andWhere('shp.student_id', studentID)
+            .andWhere('shp.status', 'ACCEPTED')
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/DAO/matchDAO.js:95]: added

File: backend/DAO/registrationDAO.js [hunk 4]
```diff
@@ -21,4 +21,17 @@ module.exports = {
 
         return result[0];
     },
+
+    getAccountByEmail: async (email) => {
+        let result = await database('registration')
+            .select()
+            .where('email', email)
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/DAO/registrationDAO.js:31]: added

File: backend/DAO/studentDAO.js [hunk 5]
```diff
@@ -243,6 +242,14 @@ module.exports = {
             });
         return result;
     },
+    
+    getStudentByUserId: async (userId) => {
+        let result = await database('student')
+            .select('id', 'name')
+            .where('user_id', userId)
+
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/DAO/studentDAO.js:250]: added

File: backend/DB/migrations/20210109014342_delete_phone_and_email_from_student.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+
+exports.up = async function(knex) {
+    let check1 = await knex.schema.hasColumn('student', "email");
+    let check2 = await knex.schema.hasColumn('student', "password");
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/DB/migrations/20210109014342_delete_phone_and_email_from_student.js:4]: fixed

File: backend/docs/documentation.yaml [hunk 7]
```diff
@@ -812,9 +929,9 @@ paths:
                       {id: 6},
                     ],
                     otherSkills: [
-                      "Cooking",
+                      "Artificial Intelligence",
                       "Organizing events",
-                      "Trekking"
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/docs/documentation.yaml:817]: updated

File: backend/routes/matchRoute.js [hunk 8]
```diff
@@ -3,16 +3,17 @@ const router = require('express').Router();
 const matchService = require('../services/matchService');
 const SuperError = require('../errors').SuperError;
 const ERR_BAD_REQUEST = require('../errors').ERR_BAD_REQUEST;
+const ERR_FORBIDDEN = require('../errors').ERR_FORBIDDEN;
 
 
 router.post("/matching", async (req, res, next) => {
-    const choice = await matchService.saveChoice(req.body);
+    const choice = await matchService.saveChoice(req.body, req.user);
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/routes/matchRoute.js]: fixed

File: backend/routes/registrationRoute.js [hunk 9]
```diff
@@ -10,4 +10,22 @@ router.post("/registration", async ( req, res, next ) => {
...
+router.post("/login", async ( req, res, next ) => {
+    let userProfile = await registrationService.getUserSession(req.body)
+        .catch(error => {
+            res.status(error.code).send(error.message);
+        });
+    res.json(userProfile);
+});
+
+router.delete("/user/:id", async (req, res, next) => {
+    // TODO: allow only logged in users to delete themselves
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/routes/registrationRoute.js]: implemented

File: backend/routes/studentRoute.js [hunk 10]
```diff
@@ -26,33 +26,25 @@ router.get("/student/:id", async (req, res, next) => {
...
-    const studentToSkills = await studentService.saveStudentProfile(req.params.id, req.body);
-    res.json(studentToSkills);
-
 });
 
-router.get("/students/last", async (req, res, next) => {
-    const lastStudent = await studentService.getLastStudent();
-    res.json(lastStudent);
+router.get("/student/account/:userID", async (req, res, next) => {
+    // TODO: THE userID variable should be the same as userID of the logged in user
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/routes/studentRoute.js]: implemented

File: backend/services/matchService.js [hunk 11]
```diff
@@ -1,8 +1,18 @@
...
 
 module.exports = {
-    saveChoice: async (choice) => { 
+    saveChoice: async (choice, auth) => { 
+        let employer = await employerService.getEmployerByPlacementId(choice.placementID);
+        if (auth.studentId !== choice.studentID && auth.employerId !== employer.id) {
+            throw new SuperError(ERR_FORBIDDEN, 'You are not authorized to save this match');
+            return;
+        }
+        //TODO: CHECK THAT IF THE ACCOUNT IS STUDENT, THE STUDENT_ACCEPT IS SET (equivalent for employer)
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/services/matchService.js]: implemented

File: backend/services/middlewareService.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
+const jwt = require('jsonwebtoken')
+
+module.exports = {
+    
+    authMiddleware: function (req, res, next) {
+        if(req.url === '/login' || req.url === '/registration' || /\/user\/[0-9]*$/.test(req.url)) {
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/services/middlewareService.js]: fixed, very good catch!

File: backend/DAO/registrationDAO.js [hunk 13]
```diff
@@ -21,4 +21,17 @@ module.exports = {
...
+        return result.length ? result[0] : null
+    },
+
+    deleteUserById: function(userId) {
+        return database('registration')
+            .where('id', userId)
+            .del();
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/DAO/registrationDAO.js:38]: fixed

File: backend/services/registrationService.js [hunk 14]
```diff
@@ -8,5 +13,55 @@ module.exports = {
...
+        return userResult;
+    },
+    updateJWT: (newInfo) => {
+        let JWT = jwt.sign(newInfo, process.env.ACCESS_TOKEN_SECRET, {
+            expiresIn: process.env.ACCESS_TOKEN_LIFE // 30 DAYS
+        });
+        return JWT;
+    },
+    deleteUser: function(userId) {
+        return registrationDAO.deleteUserById(userId);
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/services/registrationService.js]: done

File: backend/services/studentService.js [hunk 15]
```diff
@@ -89,7 +102,9 @@ self = module.exports = {
         }
         return location;
     },
-
+    getStudentByUserId: (userId) => {
+        return studentDAO.getStudentByUserId(userId);
```
[the remark this answers is not in the document]
Comment (aidadenisa) [reply backend/services/studentService.js]: done

**Review by AnnaBergamasco — APPROVED** [review]

**Review by AliceCasali — APPROVED** [review]

**Pull request merged**